From 3a5270c11950265878dea94020fa86d9a9dd337c Mon Sep 17 00:00:00 2001
From: Arjun <arjun_daasuramdass@comcast.com>
Date: Fri, 28 Oct 2022 15:12:03 +0000
Subject: [PATCH] RDKB-45261: dbus: add CCSP APIs patch for dbus 1.16.2

Reason for Change:
        When building RDK-B, the meta-rdk dbus .bbappend applies a version
        specific patch to dbus to enable support for CCSP APIs. This patch
        is the version for dbus 1.16.2 (OE 6.x / wrynose), adapted from the
        original 1.14.x patch. The build system changed from Autotools to
        Meson in dbus 1.15+, so dbus/Makefile.am is replaced by
        dbus/meson.build in this patch.

Source: COMCAST

Upstream-Status: Pending

Signed-off-by: Arjun <arjun_daasuramdass@comcast.com>

--- a/bus/activation.c
+++ b/bus/activation.c
@@ -185,7 +185,7 @@
 
   if (pending_activation->timeout_added)
     {
-      _dbus_loop_remove_timeout (bus_context_get_loop (pending_activation->activation->context),
+      dbus_loop_remove_timeout (bus_context_get_loop (pending_activation->activation->context),
                                  pending_activation->timeout);
       pending_activation->timeout_added = FALSE;
     }
@@ -1396,7 +1396,7 @@
 {
   /* FIXME use preallocated OOM messages instead of bus_wait_for_memory() */
   while (!try_send_activation_failure (pending_activation, how))
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   /* Destroy this pending activation */
   _dbus_hash_table_remove_string (pending_activation->activation->pending_activations,
@@ -1550,7 +1550,7 @@
 {
   BusPendingActivation *pending_activation = data;
 
-  return _dbus_loop_add_watch (
+  return dbus_loop_add_watch (
       bus_context_get_loop (pending_activation->activation->context),
       watch);
 }
@@ -1561,7 +1561,7 @@
 {
   BusPendingActivation *pending_activation = data;
 
-  _dbus_loop_remove_watch (bus_context_get_loop (pending_activation->activation->context),
+  dbus_loop_remove_watch (bus_context_get_loop (pending_activation->activation->context),
                            watch);
 }
 
@@ -1571,7 +1571,7 @@
 {
   BusPendingActivation *pending_activation = data;
 
-  _dbus_loop_toggle_watch (bus_context_get_loop (pending_activation->activation->context),
+  dbus_loop_toggle_watch (bus_context_get_loop (pending_activation->activation->context),
                            watch);
 }
 #endif
@@ -2030,7 +2030,7 @@
           return FALSE;
         }
 
-      if (!_dbus_loop_add_timeout (bus_context_get_loop (activation->context),
+      if (!dbus_loop_add_timeout (bus_context_get_loop (activation->context),
                                    pending_activation->timeout))
         {
           _dbus_verbose ("Failed to add timeout for pending activation\n");
--- a/bus/bus.c
+++ b/bus/bus.c
@@ -121,7 +121,7 @@
 
   context = server_get_context (server);
 
-  return _dbus_loop_add_watch (context->loop, watch);
+  return dbus_loop_add_watch (context->loop, watch);
 }
 
 static void
@@ -133,7 +133,7 @@
 
   context = server_get_context (server);
 
-  _dbus_loop_remove_watch (context->loop, watch);
+  dbus_loop_remove_watch (context->loop, watch);
 }
 
 static void
@@ -145,7 +145,7 @@
 
   context = server_get_context (server);
 
-  _dbus_loop_toggle_watch (context->loop, watch);
+  dbus_loop_toggle_watch (context->loop, watch);
 }
 
 static dbus_bool_t
@@ -157,7 +157,7 @@
 
   context = server_get_context (server);
 
-  return _dbus_loop_add_timeout (context->loop, timeout);
+  return dbus_loop_add_timeout (context->loop, timeout);
 }
 
 static void
@@ -169,7 +169,7 @@
 
   context = server_get_context (server);
 
-  _dbus_loop_remove_timeout (context->loop, timeout);
+  dbus_loop_remove_timeout (context->loop, timeout);
 }
 
 static void
@@ -809,7 +809,7 @@
       goto failed;
     }
 
-  context->loop = _dbus_loop_new ();
+  context->loop = dbus_loop_new ();
   if (context->loop == NULL)
     {
       BUS_SET_OOM (error);
@@ -1277,7 +1277,7 @@
 
       if (context->loop)
         {
-          _dbus_loop_unref (context->loop);
+          dbus_loop_unref (context->loop);
           context->loop = NULL;
         }
 
--- a/bus/connection.c
+++ b/bus/connection.c
@@ -240,7 +240,7 @@
       dbus_error_init (&error);
         
       while ((transaction = bus_transaction_new (d->connections->context)) == NULL)
-        _dbus_wait_for_memory ();
+        dbus_wait_for_memory ();
         
       if (!bus_service_remove_owner (service, connection,
                                      transaction, &error))
@@ -251,7 +251,7 @@
             {
               dbus_error_free (&error);
               bus_transaction_cancel_and_free (transaction);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto retry;
             }
           else
@@ -290,7 +290,7 @@
 
   if (d->pending_unix_fds_timeout)
     {
-      _dbus_loop_remove_timeout (bus_context_get_loop (d->connections->context),
+      dbus_loop_remove_timeout (bus_context_get_loop (d->connections->context),
                                  d->pending_unix_fds_timeout);
       _dbus_timeout_unref (d->pending_unix_fds_timeout);
     }
@@ -358,7 +358,7 @@
 {
   DBusConnection *connection = data;
 
-  return _dbus_loop_add_watch (connection_get_loop (connection), watch);
+  return dbus_loop_add_watch (connection_get_loop (connection), watch);
 }
 
 static void
@@ -367,7 +367,7 @@
 {
   DBusConnection *connection = data;
   
-  _dbus_loop_remove_watch (connection_get_loop (connection), watch);
+  dbus_loop_remove_watch (connection_get_loop (connection), watch);
 }
 
 static void
@@ -376,7 +376,7 @@
 {
   DBusConnection *connection = data;
 
-  _dbus_loop_toggle_watch (connection_get_loop (connection), watch);
+  dbus_loop_toggle_watch (connection_get_loop (connection), watch);
 }
 
 static dbus_bool_t
@@ -385,7 +385,7 @@
 {
   DBusConnection *connection = data;
   
-  return _dbus_loop_add_timeout (connection_get_loop (connection), timeout);
+  return dbus_loop_add_timeout (connection_get_loop (connection), timeout);
 }
 
 static void
@@ -394,7 +394,7 @@
 {
   DBusConnection *connection = data;
   
-  _dbus_loop_remove_timeout (connection_get_loop (connection), timeout);
+  dbus_loop_remove_timeout (connection_get_loop (connection), timeout);
 }
 
 static void
@@ -406,8 +406,8 @@
   
   if (new_status != DBUS_DISPATCH_COMPLETE)
     {
-      while (!_dbus_loop_queue_dispatch (loop, connection))
-        _dbus_wait_for_memory ();
+      while (!dbus_loop_queue_dispatch (loop, connection))
+        dbus_wait_for_memory ();
     }
 }
 
@@ -487,7 +487,7 @@
   if (connections->pending_replies == NULL)
     goto failed_4;
   
-  if (!_dbus_loop_add_timeout (bus_context_get_loop (context),
+  if (!dbus_loop_add_timeout (bus_context_get_loop (context),
                                connections->expire_timeout))
     goto failed_5;
   
@@ -562,7 +562,7 @@
 
       bus_expire_list_free (connections->pending_replies);
       
-      _dbus_loop_remove_timeout (bus_context_get_loop (connections->context),
+      dbus_loop_remove_timeout (bus_context_get_loop (connections->context),
                                  connections->expire_timeout);
       
       _dbus_timeout_unref (connections->expire_timeout);
@@ -825,7 +825,7 @@
 
   if (dbus_connection_get_dispatch_status (connection) != DBUS_DISPATCH_COMPLETE)
     {
-      if (!_dbus_loop_queue_dispatch (bus_context_get_loop (connections->context), connection))
+      if (!dbus_loop_queue_dispatch (bus_context_get_loop (connections->context), connection))
         {
           bus_dispatch_remove_connection (connection);
           goto oom;
@@ -840,7 +840,7 @@
     goto oom;
 
   _dbus_timeout_disable (d->pending_unix_fds_timeout);
-  if (!_dbus_loop_add_timeout (bus_context_get_loop (connections->context),
+  if (!dbus_loop_add_timeout (bus_context_get_loop (connections->context),
                                d->pending_unix_fds_timeout))
     goto oom;
 
--- a/bus/dir-watch-inotify.c
+++ b/bus/dir-watch-inotify.c
@@ -224,10 +224,10 @@
 
   if (watch != NULL)
     {
-      _dbus_loop_remove_watch (loop, watch);
+      dbus_loop_remove_watch (loop, watch);
       _dbus_watch_invalidate (watch);
       _dbus_watch_unref (watch);
-      _dbus_loop_unref (loop);
+      dbus_loop_unref (loop);
     }
   watch = NULL;
   loop = NULL;
@@ -266,7 +266,7 @@
       _dbus_fd_set_close_on_exec (inotify_fd);
 
       loop = bus_context_get_loop (context);
-      _dbus_loop_ref (loop);
+      dbus_loop_ref (loop);
 
       watch = _dbus_watch_new (inotify_fd, DBUS_WATCH_READABLE, TRUE,
                                _handle_inotify_watch, NULL, NULL);
@@ -278,7 +278,7 @@
           goto out;
         }
 
-      if (!_dbus_loop_add_watch (loop, watch))
+      if (!dbus_loop_add_watch (loop, watch))
         {
           bus_context_log (context, DBUS_SYSTEM_LOG_WARNING,
                            "Unable to add reload watch to main loop");
--- a/bus/dispatch.c
+++ b/bus/dispatch.c
@@ -236,7 +236,7 @@
    * until we can.
    */
   while (!bus_connection_preallocate_oom_error (connection))
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   /* Ref connection in case we disconnect it at some point in here */
   dbus_connection_ref (connection);
@@ -636,7 +636,7 @@
 {
   while (dbus_connection_get_dispatch_status (connection) ==
          DBUS_DISPATCH_NEED_MEMORY)
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   return dbus_connection_pop_message (connection);
 }
@@ -647,7 +647,7 @@
 {
   while (dbus_connection_get_dispatch_status (connection) ==
          DBUS_DISPATCH_NEED_MEMORY)
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   return dbus_connection_borrow_message (connection);
 }
@@ -763,7 +763,7 @@
           if (dbus_error_has_name (&error, DBUS_ERROR_NO_MEMORY))
             {
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto reget_service_info_data;
             }
           else
@@ -824,7 +824,7 @@
   _dbus_assert (s != NULL);
 
   while ((base_service = _dbus_strdup (s)) == NULL)
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   dbus_connection_ref (connection);
 
@@ -1057,7 +1057,7 @@
             {
               _dbus_verbose ("no memory to get service name arg from hello\n");
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto retry_get_hello_name;
             }
           else
@@ -1071,7 +1071,7 @@
       _dbus_verbose ("Got hello name: %s\n", name);
 
       while (!dbus_bus_set_unique_name (connection, name))
-        _dbus_wait_for_memory ();
+        dbus_wait_for_memory ();
 
       socd.expected_kind = SERVICE_CREATED;
       socd.expected_service_name = name;
@@ -1117,7 +1117,7 @@
             {
               _dbus_verbose ("no memory to get service name arg from acquired\n");
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto retry_get_acquired_name;
             }
           else
@@ -1368,7 +1368,7 @@
             {
               _dbus_verbose ("no memory to get uid by GetConnectionUnixUser\n");
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto retry_get_property;
             }
           else
@@ -1516,7 +1516,7 @@
             {
               _dbus_verbose ("no memory to get pid by GetConnectionUnixProcessID\n");
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto retry_get_property;
             }
           else
@@ -2110,7 +2110,7 @@
           if (dbus_error_has_name (&error, DBUS_ERROR_NO_MEMORY))
             {
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto reget_service_name_arg;
             }
           else
@@ -2215,7 +2215,7 @@
           if (dbus_error_has_name (&error, DBUS_ERROR_NO_MEMORY))
             {
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto reget_service_name_arg;
             }
           else
@@ -2368,7 +2368,7 @@
           if (dbus_error_has_name (&error, DBUS_ERROR_NO_MEMORY))
             {
               dbus_error_free (&error);
-              _dbus_wait_for_memory ();
+              dbus_wait_for_memory ();
               goto reget_service_name_arg;
             }
           else
@@ -4470,7 +4470,7 @@
 	    {
 	      _dbus_verbose ("no memory to list services by %s\n", method);
 	      dbus_error_free (&error);
-	      _dbus_wait_for_memory ();
+	      dbus_wait_for_memory ();
 	      goto retry_get_property;
 	    }
 	  else
--- a/bus/expirelist.c
+++ b/bus/expirelist.c
@@ -68,7 +68,7 @@
 
   _dbus_timeout_disable (list->timeout);
 
-  if (!_dbus_loop_add_timeout (list->loop, list->timeout))
+  if (!dbus_loop_add_timeout (list->loop, list->timeout))
     goto failed;
 
   return list;
@@ -87,7 +87,7 @@
 {
   _dbus_assert (list->items == NULL);
 
-  _dbus_loop_remove_timeout (list->loop, list->timeout);
+  dbus_loop_remove_timeout (list->loop, list->timeout);
 
   _dbus_timeout_unref (list->timeout);
 
@@ -161,7 +161,7 @@
            */
           if (!(* list->expire_func) (list, link, list->data))
             {
-              next_interval = _dbus_get_oom_wait ();
+              next_interval = dbus_get_oom_wait ();
               break;
             }
         }
@@ -337,7 +337,7 @@
   dbus_bool_t result = FALSE;
 
 
-  loop = _dbus_loop_new ();
+  loop = dbus_loop_new ();
   _dbus_assert (loop != NULL);
 
 #define EXPIRE_AFTER 100
@@ -397,7 +397,7 @@
   dbus_free (item);
   
   bus_expire_list_free (list);
-  _dbus_loop_unref (loop);
+  dbus_loop_unref (loop);
   
   result = TRUE;
 
--- a/bus/main.c
+++ b/bus/main.c
@@ -280,7 +280,7 @@
   char action = '\0';
 
   while (!_dbus_string_init (&str))
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
 
   if ((reload_pipe[RELOAD_READ_END].fd > 0) &&
       _dbus_read_socket (reload_pipe[RELOAD_READ_END], &str, 1) != 1)
@@ -330,7 +330,7 @@
         if (loop != NULL)
           {
             _dbus_daemon_report_stopping ();
-            _dbus_loop_quit (loop);
+            dbus_loop_quit (loop);
           }
       }
       break;
@@ -371,7 +371,7 @@
       exit (1);
     }
 
-  if (!_dbus_loop_add_watch (loop, watch))
+  if (!dbus_loop_add_watch (loop, watch))
     {
       _dbus_warn ("Unable to add reload watch to main loop: %s",
 		  error.message);
@@ -384,7 +384,7 @@
 static void
 close_reload_pipe (DBusWatch **watch)
 {
-    _dbus_loop_remove_watch (bus_context_get_loop (context), *watch);
+    dbus_loop_remove_watch (bus_context_get_loop (context), *watch);
     _dbus_watch_invalidate (*watch);
     _dbus_watch_unref (*watch);
     *watch = NULL;
@@ -746,7 +746,7 @@
 
   _dbus_verbose ("We are on D-Bus...\n");
   _dbus_daemon_report_ready ();
-  _dbus_loop_run (bus_context_get_loop (context));
+  dbus_loop_run (bus_context_get_loop (context));
 
   bus_context_shutdown (context);
   bus_context_unref (context);
--- a/bus/utils.c
+++ b/bus/utils.c
@@ -44,7 +44,7 @@
   DBusDispatchStatus status;
 
   while ((status = dbus_connection_dispatch (connection)) == DBUS_DISPATCH_NEED_MEMORY)
-    _dbus_wait_for_memory ();
+    dbus_wait_for_memory ();
   
   return status == DBUS_DISPATCH_DATA_REMAINS;
 }
--- a/dbus/dbus-ccsp-apis.c
+++ b/dbus/dbus-ccsp-apis.c
@@ -0,0 +1,147 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/* dbus-ccsp-apis.c
+ *
+ * Copyright (C) 2003  Red Hat, Inc.
+ *
+ * Licensed under the Academic Free License version 2.1
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#include <config.h>
+#define DBUS_COMPILATION  /* building internal dbus file */
+#include <dbus/dbus-ccsp-apis.h>
+#include <dbus/dbus-internals.h>
+#include <dbus/dbus-watch.h>
+#include "dbus-mainloop.h"
+#include "dbus-connection-internal.h"
+#ifdef DBUS_UNIX
+#include <dbus/dbus-sysdeps-unix.h>
+#endif
+
+#include <unistd.h> /* for read() */
+
+/* Handle the loop wakeup watch */
+dbus_bool_t wake_handler(DBusWatch *watch, unsigned int flags, void *data)
+{
+    char ch;
+    int fd = (int)data;
+    /* Read and discard */
+    while(read(fd, &ch, 1) == 1);
+}
+
+/* We only need 1 wake watch per mainloop process */
+DBusWatch *wake_watch = NULL;
+
+/* From dbus-mainloop */
+DBusLoop* dbus_loop_new(void)
+{
+    return _dbus_loop_new();
+}
+
+DBusLoop* dbus_loop_ref(DBusLoop *loop)
+{
+    return _dbus_loop_ref(loop);
+}
+
+void dbus_loop_unref(DBusLoop *loop) { _dbus_loop_unref(loop); }
+
+dbus_bool_t dbus_loop_add_watch(DBusLoop *loop, DBusWatch *watch)
+{
+    return _dbus_loop_add_watch(loop, watch);
+}
+
+void dbus_loop_remove_watch(DBusLoop *loop, DBusWatch *watch)
+{
+    _dbus_loop_remove_watch(loop, watch);
+}
+
+void dbus_loop_toggle_watch(DBusLoop *loop, DBusWatch *watch)
+{
+    _dbus_loop_toggle_watch(loop, watch);
+}
+
+dbus_bool_t dbus_loop_add_timeout(DBusLoop *loop, DBusTimeout *timeout)
+{
+    return _dbus_loop_add_timeout(loop, timeout);
+}
+
+void dbus_loop_remove_timeout(DBusLoop *loop, DBusTimeout *timeout)
+{
+     _dbus_loop_remove_timeout(loop, timeout);
+}
+
+dbus_bool_t dbus_loop_queue_dispatch(DBusLoop *loop, DBusConnection *connection)
+{
+    return _dbus_loop_queue_dispatch(loop, connection);
+}
+
+void dbus_loop_run(DBusLoop *loop)
+{
+     _dbus_loop_run(loop);
+}
+
+void dbus_loop_quit(DBusLoop *loop)
+{
+     _dbus_loop_quit(loop);
+}
+
+dbus_bool_t dbus_loop_iterate(DBusLoop *loop, dbus_bool_t block)
+{
+    return _dbus_loop_iterate(loop, block);
+}
+
+dbus_bool_t dbus_loop_dispatch(DBusLoop *loop)
+{
+    return _dbus_loop_dispatch(loop);
+}
+
+void dbus_wait_for_memory(void)
+{
+    _dbus_wait_for_memory();
+}
+
+/* From dbus-connection */
+void dbus_connection_lock(DBusConnection *connection)
+{
+    _dbus_connection_lock(connection);
+}
+
+void dbus_connection_unlock(DBusConnection *connection)
+{
+    _dbus_connection_unlock(connection);
+}
+
+int dbus_get_oom_wait (void)
+{
+     return _dbus_get_oom_wait();
+}
+
+dbus_bool_t dbus_loop_add_wake(DBusLoop *loop, int fd)
+{
+     wake_watch = _dbus_watch_new(fd, DBUS_WATCH_READABLE, TRUE, wake_handler, (void *)fd, NULL);
+     return dbus_loop_add_watch(loop, wake_watch);
+}
+
+void dbus_loop_remove_wake(DBusLoop *loop)
+{
+     if (wake_watch)
+     {
+         dbus_loop_remove_watch(loop, wake_watch);
+         _dbus_watch_unref(wake_watch);
+         wake_watch = NULL;
+     }
+}
--- a/dbus/dbus-ccsp-apis.h
+++ b/dbus/dbus-ccsp-apis.h
@@ -0,0 +1,111 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*- */
+/* dbus-ccsp-apis.h
+ *
+ * Copyright (C) 2003  Red Hat, Inc.
+ *
+ * Licensed under the Academic Free License version 2.1
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+#if !defined (DBUS_INSIDE_DBUS_H) && !defined (DBUS_COMPILATION)
+#error "Only <dbus/dbus.h> can be included directly, this file may disappear or change contents."
+#endif
+
+#ifndef DBUS_CCSP_APIS_H
+#define DBUS_CCSP_APIS_H
+
+#include <dbus/dbus.h>
+#include <stdarg.h>
+
+typedef struct DBusLoop DBusLoop;
+
+DBUS_BEGIN_DECLS
+
+/**
+ * @addtogroup DBusCcspApis
+ * @{
+ */
+
+/* From dbus-mainloop */
+DBUS_EXPORT
+DBusLoop*   dbus_loop_new            (void);
+
+DBUS_EXPORT
+DBusLoop*   dbus_loop_ref            (DBusLoop            *loop);
+
+DBUS_EXPORT
+void        dbus_loop_unref          (DBusLoop            *loop);
+
+DBUS_EXPORT
+dbus_bool_t dbus_loop_add_watch      (DBusLoop            *loop,
+                                      DBusWatch           *watch);
+
+DBUS_EXPORT
+void        dbus_loop_remove_watch   (DBusLoop            *loop,
+                                      DBusWatch           *watch);
+DBUS_EXPORT
+void        dbus_loop_toggle_watch   (DBusLoop            *loop,
+                                      DBusWatch           *watch);
+DBUS_EXPORT
+dbus_bool_t dbus_loop_add_timeout    (DBusLoop            *loop,
+                                      DBusTimeout         *timeout);
+
+DBUS_EXPORT
+void        dbus_loop_remove_timeout (DBusLoop            *loop,
+                                      DBusTimeout         *timeout);
+
+DBUS_EXPORT
+dbus_bool_t dbus_loop_queue_dispatch (DBusLoop            *loop,
+                                      DBusConnection      *connection);
+
+DBUS_EXPORT
+void        dbus_loop_run            (DBusLoop            *loop);
+
+DBUS_EXPORT
+void        dbus_loop_quit           (DBusLoop            *loop);
+
+DBUS_EXPORT
+dbus_bool_t dbus_loop_iterate        (DBusLoop            *loop,
+                                      dbus_bool_t          block);
+
+DBUS_EXPORT
+dbus_bool_t dbus_loop_dispatch       (DBusLoop            *loop);
+
+DBUS_EXPORT
+void dbus_wait_for_memory(void);
+
+DBUS_EXPORT
+int dbus_get_oom_wait(void);
+
+/* From dbus-connection */
+DBUS_EXPORT
+void dbus_connection_lock(DBusConnection *connection);
+
+DBUS_EXPORT
+void dbus_connection_unlock(DBusConnection *connection);
+
+DBUS_EXPORT
+dbus_bool_t dbus_loop_add_wake       (DBusLoop            *loop,
+                                      int                 fd);
+
+DBUS_EXPORT
+void dbus_loop_remove_wake           (DBusLoop            *loop);
+
+/** @} */
+
+DBUS_END_DECLS
+
+#endif /* DBUS_CCSP_APIS_H */
--- a/dbus/dbus.h
+++ b/dbus/dbus.h
@@ -31,6 +31,7 @@
 #include <dbus/dbus-arch-deps.h>
 #include <dbus/dbus-address.h>
 #include <dbus/dbus-bus.h>
+#include <dbus/dbus-ccsp-apis.h>
 #include <dbus/dbus-connection.h>
 #include <dbus/dbus-errors.h>
 #include <dbus/dbus-macros.h>
--- a/dbus/meson.build
+++ b/dbus/meson.build
@@ -33,6 +33,7 @@
     'dbus-address.c',
     'dbus-auth.c',
     'dbus-bus.c',
+    'dbus-ccsp-apis.c',
     'dbus-connection.c',
     'dbus-credentials.c',
     'dbus-errors.c',
@@ -71,10 +72,13 @@
     'dbus-hash.c',
     'dbus-internals.c',
     'dbus-list.c',
+    'dbus-mainloop.c',
     'dbus-marshal-basic.c',
     'dbus-memory.c',
     'dbus-mempool.c',
     'dbus-pipe.c',
+    'dbus-pollable-set-poll.c',
+    'dbus-pollable-set.c',
     'dbus-string.c',
     'dbus-sysdeps.c',
 ]
@@ -92,10 +96,7 @@
 
 dbus_util_sources = [
     'dbus-asv-util.c',
-    'dbus-mainloop.c',
     'dbus-message-util.c',
-    'dbus-pollable-set-poll.c',
-    'dbus-pollable-set.c',
     'dbus-shell.c',
     'dbus-string-util.c',
     'dbus-sysdeps-util.c',
@@ -168,7 +169,7 @@
 
 
 if use_linux_epoll
-    dbus_util_sources += 'dbus-pollable-set-epoll.c'
+    dbus_shared_sources += 'dbus-pollable-set-epoll.c'
 endif
 
 version_script = configure_file(
@@ -229,6 +230,7 @@
 install_headers(
     'dbus-address.h',
     'dbus-bus.h',
+    'dbus-ccsp-apis.h',
     'dbus-connection.h',
     'dbus-errors.h',
     'dbus-macros.h',
